client/state/utils.js

```diff
@@ -442,6 +453,61 @@ function serializeState( reducers, state, action ) {
 	);
 }
 
+/**
+ * Create a new reducer from original `reducers` by adding a new `reducer` at `keyPath`
+ * @param {Object} reducers Object with reducer names as keys and reducer functions as values that
+ *   is used as parameter to `combineReducers` (the original Redux one and our extension, too).
+ * @return {Function} The function to be attached as `addReducer` method to the
+ *   result of `combineReducers`.
+ */
+export function addReducer( reducers ) {
+	return ( keyPath, reducer ) => {
+		// extract the first key from keyPath and dive recursively into the reducer tree
+		const [ key, ...restKeys ] = keyPath;
+
+		const existingReducer = reducers[ key ];
+		let newReducer;
+
+		// if there is an existing reducer at this path, we'll recursively call `addReducer`
+		// until we reach the final destination in the tree.
+		if ( existingReducer ) {
+			// we reached the final destination in the tree and another reducer already lives there!
+			if ( restKeys.length === 0 ) {
+				throw new Error( `Reducer with key '${ key }' is already registered` );
+			}
+
+			if ( ! existingReducer.addReducer ) {
+				throw new Error(
+					"New reducer can be added only into a reducer created with 'combineReducers'"
+				);
+			}
+
+			newReducer = existingReducer.addReducer( restKeys, reducer );
+		} else {
+			// for the remaining keys in the keyPath, create a nested reducer:
+			// if `restKeys` is `[ 'a', 'b', 'c']`, then the result of this `reduceRight` is:
+			// ```js
+			// combineReducers( {
+			//   a: combineReducers ( {
+			//     b: combineReducers( {
+			//       c: reducer
+			//     } )
+			//   })
+			// })
+			// ```
+			newReducer = reduceRight(
+				restKeys,
+				( subreducer, subkey ) => createCombinedReducer( { [ subkey ]: subreducer } ),
+				setupReducerPersistence( reducer )
```

Comment:
Do we need any tests to verify we are handling persistence properly for newly added reducers?

Reply:
Yes, that's the topic of one of followup PRs. Persistence of dynamic reducers really makes sense only after the state is serialized into multiple IndexedDB rows, according to `storageKey`. See also #28534 and the whole story in #27415 and https://github.com/Automattic/wp-calypso/pull/27415/commits/99c4816b7f7777dc5cd97bb8632978d927a358e4 (the link to this commit will likely expire soon after the next rebase)